lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -441,6 +1975,244 @@ bool StackAnalysis::runOnModule(Module &M) {
+                                          llvm::GlobalValue::ExternalLinkage,
+                                          "postcall_hook",
+                                          M);
+  PostHookMarker->addFnAttr(llvm::Attribute::ReadOnly);
+  PostHookMarker->addFnAttr(llvm::Attribute::NoUnwind);
+
+  auto *RetHookMarker = Function::Create(FTy,
+                                         llvm::GlobalValue::ExternalLinkage,
+                                         "retcall_hook",
+                                         M);
+  RetHookMarker->addFnAttr(llvm::Attribute::ReadOnly);
+  RetHookMarker->addFnAttr(llvm::Attribute::NoUnwind);
+
+  // Initialize marker to adjust `unexpectedpc` basic block for fake functions
+  FTy = llvm::FunctionType::get(llvm::Type::getVoidTy(M.getContext()), false);
+  auto
+    *UnexpectedPCMarker = Function::Create(FTy,
+                                           llvm::GlobalValue::ExternalLinkage,
+                                           "unexpectedpc_hook",
+                                           M);
+  UnexpectedPCMarker->addFnAttr(llvm::Attribute::ReadOnly);
+  UnexpectedPCMarker->addFnAttr(llvm::Attribute::NoUnwind);
+
+  // Queue to be populated with the CFEP
+  llvm::SmallVector<BasicBlock *, 8> Worklist;
+  llvm::SmallVector<BasicBlockNode *, 4> EntryBlocks;
+  UniquedQueue<BasicBlockNode *> CFEPQueue;
+  SmallCallGraph CG{};
+
+  using BasicBlockToNodeMapTy = llvm::DenseMap<BasicBlock *, BasicBlockNode *>;
+  BasicBlockToNodeMapTy BasicBlockNodeMap;
+
+  // Create over-approximated call graph
+  for (CFEP &Function : Functions) {
+    BasicBlockNode Node{ Function.Entry };
+    auto *GraphNode = CG.addNode(Node);
+    BasicBlockNodeMap[Function.Entry] = GraphNode;
+  }
+
+  for (CFEP &Function : Functions) {
+    llvm::SmallSet<BasicBlock *, 8> Visited;
+    auto *StartNode = BasicBlockNodeMap[Function.Entry];
+    Worklist.emplace_back(Function.Entry);
+
+    do {
+      auto *Current = Worklist.pop_back_val();
+      Visited.insert(Current);
+      for (auto *Successor : successors(Current)) {
+        if (isFunctionCall(Current)) {
+          auto *Callee = getFunctionCallCallee(Current);
+          // If not an indirect call, add the node to the CG
+          if (Callee) {
+            auto *Node = BasicBlockNodeMap[Callee];
+            StartNode->addSuccessor(Node);
+          }
+          auto *Next = getFallthrough(Current);
+          if (!Visited.count(Next))
+            Worklist.push_back(Next);
+          continue;
+        }
+
+        if (!GCBI::isPartOfRootDispatcher(Successor)
+            && !Visited.count(Successor))
+          Worklist.push_back(Successor);
+      }
+    } while (!Worklist.empty());
+  }
+
+  for (const auto &[_, Node] : BasicBlockNodeMap) {
+    if (!Node->hasPredecessors())
+      EntryBlocks.emplace_back(Node);
+
+    // Recursive?
+    if (*Node->predecessors().begin() == Node)
+      EntryBlocks.emplace_back(Node);
+  }
```

Comment:
This is suspicious. You're not considering indirect recursion. You're not considering `A -> B -> A`.
What are you trying to achieve? I guess you're trying to find all the "entry points" of the call graph. Problem is: you might have loops with no obvious entry points.

One thing you can do is use `scc_begin`/`scc_end` and, if every node in the SCC has a predecessors, just mark a random node of the SCC as an entry node.

`scc_begin`/`scc_end` is a bit delicate, IIRC each node on itself represents a SCC. But I might be wrong, play around with it a bit.